awx/api/validators.py

```diff
@@ -0,0 +1,64 @@
+from django.core.validators import RegexValidator, validate_ipv46_address
+from django.core.exceptions import ValidationError
+
+
+class HostnameRegexValidator(RegexValidator):
+    """
+    Fully validates a domain name that is compliant with norms in Linux/RHEL
+        - Cannot start with a hypen
+        - Cannot being with, or end with a "."
+        - Cannot contain any whitespaces
+        - Entire hostname is max 255 chars (including dots)
+        - Each domain/label is between 1 and 63 characters, except top level domain, which must be at least 2 characters
+        - Supports ipv4, ipv6, simple hostnames and FQDNs
+        - Follows RFC 9210 (modern RFC 1123, 1178) requirements
+
+    Excepts an IP Address or Hostname as the argument
+    """
+
+    def __call__(self, value):
+        regex_matches, err = self.__validate(value)
+        invalid_input = regex_matches if self.inverse_match else not regex_matches
+        if invalid_input:
+            if err is None:
+                err = ValidationError(self.message, code=self.code, params={"value": value})
+            raise err
+
+    def __str__(self):
+        return f"regex={self.regex}, message={self.message}, code={self.code}, inverse_match={self.inverse_match}, flags={self.flags}"
+
+    def __validate(self, value):
+
+        if len(value.split(" ")) > 1:
+            return False, ValidationError("whitespaces in hostnames are illegal")
+
+        """
+        If we have an IP address, try and validate it.
+        """
+        is_ip_addr = False
+        try:
+            validate_ipv46_address(value)
+            is_ip_addr = True
+        except ValidationError:
+            pass
+
+        if is_ip_addr:
+            if self.regex.match(value):
+                return True, None
+
+        """
+        By this point in the code, we probably have a simple hostname, FQDN or a strange hostname like "192.localhost.domain.101"
+        """
+        if len(value) > 255:
+            return False, ValidationError("hostname cannot be longer than 255 characters")
+
+        for i, label in enumerate(value.split(".")):
```

Comment:
I feel like most if not all of these conditions can be captured by the regex.  Off the top of my head, something like `([a-z0-9][-a-z0-9]{0,62}[.])*[a-z0-9][-a-z0-9]{1,}`.  Including the exclusion of spaces up on lines 32-33.